File: LeetCode_Algorithm/200/200.py
from typing import List

class Solution:
    # 515ms 26%
    def numIslands(self, grid: List[List[str]]) -> int:
        ans = 0
        for c in range(len(grid[0])):
            for r in range(len(grid)):
                if grid[r][c] == '1':
                    ans += 1
                    self.solve(grid, r, c)
        return ans

    def solve(self, grid, r, c):
        dr = [0, 1, 0, -1]
        dc = [1, 0, -1, 0]
        grid[r][c] = '0'
        for i in range(4):
            r_idx = r + dr[i]
            c_idx = c + dc[i]
            if 0 <= r_idx < len(grid) and 0 <= c_idx < len(grid[0]):
                if grid[r_idx][c_idx] == '1':
                    self.solve(grid, r_idx, c_idx)

print(Solution().numIslands([
  ["1","1","1","1","0"],
  ["1","1","0","1","0"],
  ["1","1","0","0","0"],
  ["0","0","0","0","1"]
]))

File: LeetCode_Algorithm/200/200.js
function check(visited, r, c) {
  for (let arr of visited) {
    if (arr[0] == r && arr[1] == c){
      return false
    }
  }
  return true
}
function solve(grid, r, c, visited) {
  // O(n) O(n) 5432ms 5%
  grid[r][c] = '0'
  dr = [0, 1, 0, -1]
  dc = [1, 0, -1, 0]

  for (let i=0; i < 4 ; i++) {
    let newR = r+dr[i]; let newC = c+dc[i]
    if (0 <= newR && newR < grid.length && 0 <= newC && newC < grid[0].length) {
      if(grid[newR][newC] == '1' && check(visited, newR, newC)){
        solve(grid, newR, newC, visited)
      }
      visited.push(newR, newC)
    }
  }
}
var numIslands = function(grid) {
  var ans = 0
  var visited = []
  for (let c in grid[0]) {
    for (let r in grid) {
      if (grid[r][c] == "1") {
        ans += 1
        solve(grid, parseInt(r), parseInt(c), visited)
      }
    }
  }
  return ans
};

console.log(numIslands(
  [["1","1","1","1","0"],
    ["1","1","0","1","0"],
    ["1","1","0","0","0"],
    ["0","0","0","0","0"]]))
